### PR TITLE
Add icu-locale macros

### LINKED ISSUE #75: Add macros for locale/langid
One of the features of `unic-locale` is that it allows for "free" encoding of language identifiers, locales and subtags thanks to proc macros.

In the current model it's quite quirky, and I didn't want to include it in the initial landing, but for that code to work, we only need to add two methods per subtag, and at least one of them doesn't have to be public - they just need to make it possible to create a subtag from a pre-computed u64/u32 (which is what the proc macro will do).

The good news is that Rust is about to [stabilize proc macros](https://github.com/rust-lang/rust/pull/68717) (targeting Rust 1.45) which will allow us to get this feature without multi-crate hacks required before.

### PR DESCRIPTION
**Pull request by zbraniecki**
Fixes #75.

### CODE REVIEW

**Comment by zbraniecki** [comment]
@echeran I may need your help in getting the actions cache invalidated - the failures are because we use outdated Rust. We should bump to 1.46 which will happen on cache regeneration.

The rest is ready for review!

I had to wrap `Variants` in `Option` to allow it to be empty from `const fn` so that `langid!` macro works as const. I think it's a small complication that allows for a major usability improvement and it is an internal change that can be safely cleaned up once Rust gains support for `Box::new([])` in `const fn`.

The rest is fairly simple I hope :)

**Comment by echeran** [comment]
I created an issue to separate out the discussion of the cached artifacts causing conflicts - #225

File: components/locale/macros/src/lib.rs
Comment (sffc) [inline components/locale/macros/src/lib.rs:36]: Here and below: I think macros aren't namespaced.  `language!` seems awfully generic.  Consider `icu_language!`?  I did `icu_data_key!` for my macro in DataProvider.

File: components/locale/macros/src/lib.rs
Comment (sffc) [inline components/locale/macros/src/lib.rs:47]: These macros are public and should have rustdoc, right?

File: components/locale/src/subtags/language.rs
Comment (sffc) [inline components/locale/src/subtags/language.rs:85]: Here and elsewhere: It seems odd to return a `u64`.  Returning the actual `TinyStr8` would have more type safety, and then you can call another method to convert the `TinyStr8` into the raw untyped value if you so desire.

If you make this change, then `from_raw_unchecked` should also be changed to take a `TinyStr8`.

Consider naming the functions `into_tinystr()` and `from_tinystr_unchecked()`.

I think this would be better because it emphasizes that there are two lossy/unsafe operations:

1. Language subtag <-> TinyStr8
2. TinyStr8 <-> u64

All language subtags are TinyStr8s, and all TinyStr8s are u64s, but the reverse is not true in either case.

File: components/locale/src/subtags/variants.rs
Comment (sffc) [inline components/locale/src/subtags/variants.rs:29]: Why didn't you make the `variants` field of `LanguageIdentifier` be `Option<Variants>` like the region and script subtags?  Why did you hide the `Option` down here?  Doing this would avoid the need to make any changes to this file.

File: components/locale/macros/src/lib.rs
Comment (sffc) [inline components/locale/macros/src/lib.rs:97]: I guess `vec![]` is `const`?

It seems like if there are no variants, you should be able to pass `None` here, like you do with the script and region.

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/locale/macros/src/lib.rs
Comment (zbraniecki) [reply components/locale/macros/src/lib.rs:36]: I don't see a reason to prefix the macro names with `icu_` - I don't see a real chance that someone will have imported multiple macros named `language` into the same module.

If you feel strongly about this, I'll open an issue.

File: components/locale/src/subtags/language.rs
Comment (zbraniecki) [reply components/locale/src/subtags/language.rs:85]: @Manishearth - I remember when we worked on the macros originally for `unic-locale`, you recommended rqw `u64/u32` for raw. Do you have an opinion on switching it to `TinyStr` for reasons laid out by @sffc?

My primary consideration is whether its ok to *expose* the `TinyStr` (which so far has been an internal detail) and stabilize around it even in the "non-meant-for-public-consumption" case of from/to_raw.

WDYT?

File: components/locale/src/subtags/variants.rs
Comment (zbraniecki) [reply components/locale/src/subtags/variants.rs:29]: Mostly because logically there's no value in `Option` of collection since it doesn't have any separate meaning from an empty collection.

For that reason, my target is to have `variants` field be a collection, and in `Variants` I had to re-add the `Option` just because of the limitations of const function.
Since fields on `LanguageIdentifier` are public, having `Option` there makes it a public API, while having the `Option` here is an internal (except of from/to_raw) and I don't expect any customer to write logic which involves handling `variants` as an option of a collection.

File: components/locale/macros/src/lib.rs
Comment (zbraniecki) [reply components/locale/macros/src/lib.rs:97]: explained my logic in the comment above.

File: components/locale/macros/src/lib.rs
Comment (zbraniecki) [reply components/locale/macros/src/lib.rs:103]: Yes! Added!

**Comment by zbraniecki** [comment]
I rebased on top of master, added doc comments and improved handling of variants in macro.

This PR is pending release of Rust 1.47 on October 8th (https://forge.rust-lang.org/), but I'd like to get it reviewed and ready for landing prior to that.

File: components/locale/macros/src/lib.rs
Comment (zbraniecki) [reply components/locale/macros/src/lib.rs:72]: Filed https://github.com/unicode-org/icu4x/issues/253

File: components/locale/src/subtags/language.rs
Comment (Manishearth) [reply components/locale/src/subtags/language.rs:85]: I think if we are exposing the raw data in the public API, we are exposing the TinyStr _format_ even if not the TinyStr _type_, so we're halfway there, anyway. I think it makes more sense to return/accept TinyStr here.

File: components/locale/src/subtags/variants.rs
Comment (sffc) [reply components/locale/src/subtags/variants.rs:29]: Ok.  Let's continue the discussion in #243

File: components/locale/macros/src/lib.rs
Comment (sffc) [inline components/locale/macros/src/lib.rs:53]: Here and elsewhere (this might be a pre-existing issue): I would avoid using the term "perform canonicalization".  UTS 35 defines two types of canonicalization:

- *canonical syntax*
- *canonical form*

You are doing the first but not the second (#218).

Alternative language options:

- The macro will return the tag in canonical syntax.
- The macro will transform the tag into canonical syntax.
- The macro will perform syntax canonicalization of the tag.  (Not as precise, because "syntax canonicalization" is not a term that you can find in UTS 35.)

http://unicode.org/reports/tr35/#Canonical_Unicode_Locale_Identifiers

**Comment by zbraniecki** [comment]
> Here and elsewhere (this might be a pre-existing issue): I would avoid using the term "perform canonicalization". UTS 35 defines two types of canonicalization:

I applied your feedback to docs, except of Locale and LanguageIdentifier `canonicalize` method doc which already stated that we perform "all available levels of canonicalization".

I generally believe that cases where anyone cares about some syntax canonicalization but not form are so extremely rare, that I would like to introduce form canonicalization as an implicit "upgrade" to the functionality without changes to the signature and considered non-breaking.

Since it requires data, we may end up having it behind a feature, in which case user can in theory use that feature to decide if the canonicalization will be syntax only or also form, but I'd like to treat the public API as "we will do everything we can do give you a good output" and if we see a body of need accruing for more specificity, we can then break the API and add options to flip parts of canonicalization.

I updated the docs but I'd like to remove the specifier "syntax" once we land the form canonicalization.

Either way, that's a conversation we can have when we work on form canonicalization and I hope this update unblocks this PR.

File: components/locale/macros/src/lib.rs
Comment (sffc) [reply components/locale/macros/src/lib.rs]: Zibi's reply: https://github.com/unicode-org/icu4x/pull/220#issuecomment-695356372

File: components/locale/macros/src/token_stream.rs
Comment (sffc) [inline components/locale/macros/src/token_stream.rs:9]: Why `self` instead of `&self`?

File: components/locale/macros/src/token_stream.rs
Comment (sffc) [inline components/locale/macros/src/token_stream.rs:12]: I don't see why Self needs to be Sized?

**Review by sffc — DISMISSED** [review]

File: components/locale/src/subtags/language.rs
Comment (zbraniecki) [reply components/locale/src/subtags/language.rs:85]: In my experience `into|from`_raw are used to describe code that exposes inner value (usually for some form of serialization) and accepts it back to recreate the struct from that serialization.
https://doc.rust-lang.org/std/option/enum.Option.html?search=into_raw
https://doc.rust-lang.org/std/option/enum.Option.html?search=from_raw

File: components/locale/macros/src/token_stream.rs
Comment (zbraniecki) [reply components/locale/macros/src/token_stream.rs:9]: To stay aligned with `Into` convention which consumes the source when producing the output - https://doc.rust-lang.org/std/convert/trait.Into.html#tymethod.into

It's a helper function, used only during compilation.

File: components/locale/macros/src/token_stream.rs
Comment (zbraniecki) [reply components/locale/macros/src/token_stream.rs:12]: Without it:

```
error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> components/locale/macros/src/token_stream.rs:10:26
   |
10 |     fn into_token_stream(self) -> Result<TokenStream, proc_macro::LexError>
   |                          ^^^^ doesn't have a size known at compile-time
   |
   = help: unsized locals are gated as an unstable feature
help: consider further restricting `Self`
   |
10 |     fn into_token_stream(self) -> Result<TokenStream, proc_macro::LexError> where Self: Sized
   |                                                                             ^^^^^^^^^^^^^^^^^
help: function arguments must have a statically known size, borrowed types always have a known size
   |
10 |     fn into_token_stream(&self) -> Result<TokenStream, proc_macro::LexError>
   |                          ^

error: aborting due to previous error
```

**Review by nciric — APPROVED** [review]

File: components/locale/macros/src/lib.rs
Comment (sffc) [inline components/locale/macros/src/lib.rs:8]: Looking for literal `"` is also what you do in tinystr.  We need to fix this, but it shouldn't block this PR.  https://github.com/zbraniecki/tinystr/issues/24

**Review by sffc — APPROVED** [review]

**Pull request merged**
